### PR TITLE
Updated README and added code for change code

### PR DESCRIPTION
**Pull request by cyrusHwang**
@AlexEdgcomb Ready for review.

### CODE REVIEW

File: lib/change_code.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                        user.setHighestLScore(lScore)
+                        user.setMostChangedSub(rawSub)
+        sortedData = sorted(userIdToUser.items(), key = lambda record : record[1].highestLScore, reverse = True)
+        for userID, user in sortedData:
+            writer.writerow([userID, len(user.uqSubs), user.mostChangedSub])
+        totNumUsers = len(allUsers)
+        numParticipators = len(userIdToUser)
+        print("Percentage of students that changed initial code for {}: {:.1%}".format(contentID, numParticipators/totNumUsers))
+        print("Average number of unique, altered submissions for {}: {:.2f}\n".format(contentID, sum([len(user.uqSubs) for user in userIdToUser.values()])/numParticipators))
+
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: Empty lines can be removed

File: lib/change_code.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import csv
+from collections import defaultdict
+from lib import misc_funcs
+import html
+import urllib.request
+import json
+
+class User:
+    def __init__(self):
+        self.uqSubs = set()
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: You might checkout out the official style standard:
https://www.python.org/dev/peps/pep-0008/

File: lib/change_code.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import csv
+from collections import defaultdict
+from lib import misc_funcs
+import html
+import urllib.request
+import json
+
+class User:
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: This class might be defined in a separate file and imported to this file

File: lib/change_code.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    for contentID, path in ogCsvData["cIdToPath"].items():
+        outputCsv = open("generated_csvs/{}/change_code/{}_change_answer_{}.csv".format(zyBookCode, zyBookCode, contentID),
+                         "w", newline='', encoding="utf8")
+        writer = csv.writer(outputCsv)
+        writer.writerow(["User ID", "Number of Unique, Altered Subs", "Most Altered Sub"])
+        
+        responseURL = "http://s2.zybooks.com/v1/content_resource/{}".format(contentID)
+        responseContentText = urllib.request.urlopen(responseURL)
+        responseContentObj = json.loads(responseContentText.read().decode("utf-8-sig"))
+        default = html.unescape(responseContentObj["content_resource"]["payload"]["options"]["htmlContents"]).strip()
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: `default` could use a more descriptive name. `default_???`

File: lib/misc_funcs.py [hunk 5]
```diff
@@ -23,8 +23,33 @@ def stripNoiseSub(string : str):
...
+
+def levenshtein(s1, s2):
+    '''
+    Calculates Levenshtein score for two strings: the higher it is the more different they are from each other.
+    A score of 0 means they are the same.
+    '''
+    if len(s1) < len(s2):
+        return levenshtein(s2, s1)
+
+    # len(s1) >= len(s2)
```
Comment (AlexEdgcomb) [inline lib/misc_funcs.py]: Remove this comment?

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , I've made a pass. I like how you're building code blocks. You might include a comment describing in plain English what each block is doing. Ex: `# Download student activity data`

**Comment by cyrusHwang** [comment]
@AlexEdgcomb ready for review *haven't done vagrant yet, but made some other changes.

File: DataAnalysis.py [hunk 6]
```diff
@@ -9,63 +9,83 @@
...
+1) manual entry of content IDs
+2) rIDs text file
+3) manual entry of chapter[.lesson]s
+4) all activities
+Input a number: ''').strip()
+
+misc_funcs.makeDirIfNotExist("original_csvs/{}".format(zyBookCode))
+if choice == '1':
+    contentIDsList = input("Enter content IDs separated by spaces: ").strip().split(' ')
+    print("Downloading csvs...")
```
Comment (AlexEdgcomb) [inline DataAnalysis.py]: Refactoring opportunity: This line of code is repeated a few times below. Looks like you can simply move this line before the if-else's.

File: lib/activity.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#definition of Activity class
```
Comment (AlexEdgcomb) [inline lib/activity.py]: This comment is pretty bland. Please explain what an Activity class is. Also, please example what each internal variable is used for.

File: lib/activity.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+
+class Activity:
+    def __init__(self):
+        self.numTotAtts = 0
```
Comment (AlexEdgcomb) [inline lib/activity.py]: Please write out the full variable name. No need to be cryptic. You're computer can store the extra characters :)

Also, Python variable naming is uses underscores instead of camelcase, ex: `number_of_total_attempts`

File: lib/activity.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+        self.numUqAtts = 0
```
Comment (AlexEdgcomb) [inline lib/activity.py]: Same comment as above and below :)

Comment (AlexEdgcomb) [inline lib/activity.py]: Btw, here's the Python style guide:
https://www.python.org/dev/peps/pep-0008/

File: lib/activity_difficulty.py [hunk 10]
```diff
@@ -1,25 +1,28 @@
...
     
     Content ID,  Median Number of Unique Attempts, Average Number of Unique Attempts, Median Number of Total Attempts,
     Average Number of Total Attempts, Number of Users Participated
 
     Records are sorted in descending order by median number of unique attempts.
     '''
-    cIdToAttsData = defaultdict(dict)
+    cIdToActivity = defaultdict(dict)
     zyBookCode = ogCsvData["zyBookCode"]
     for contentID, path in ogCsvData["cIdToPath"].items():
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Please briefly comment what this loop is doing

File: lib/activity_difficulty.py [hunk 11]
```diff
@@ -28,24 +31,23 @@ def generateData(ogCsvData):
...
-            cIdToAttsData[contentID]["medianTotAtts"] = median(numAtts["lstNumTotAtts"])
-            cIdToAttsData[contentID]["avgTotAtts"] = round(mean(numAtts["lstNumTotAtts"]), 2)
-            cIdToAttsData[contentID]["numUsersPart"] = numUsersParticipated
-    sortedData = sorted(cIdToAttsData.items(), key = lambda entry : entry[1]["medianUqAtts"], reverse=True)
-    outputCsv = open("generated_csvs/{}/{}_activity_difficulty.csv".format(zyBookCode, zyBookCode),
+                    userIDtoUser[userID].numTotAtts += 1
+                    userIDtoUser[userID].uqSubs.add(misc_funcs.stripNoiseSub(rawSub))
+        currActivity.numUsersPart = len(userIDtoUser)
+        if currActivity.numUsersPart != 0:   #to avoid div by zero in calc of avg num attempts and bc we don't care about these activities
+            for user in userIDtoUser.values():
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Please briefly comment what this loop is doing

File: lib/activity_difficulty.py [hunk 12]
```diff
@@ -28,24 +31,23 @@ def generateData(ogCsvData):
...
-            cIdToAttsData[contentID]["avgUqAtts"] = round(mean(numAtts["lstNumUqAtts"]), 2)
-            cIdToAttsData[contentID]["medianTotAtts"] = median(numAtts["lstNumTotAtts"])
-            cIdToAttsData[contentID]["avgTotAtts"] = round(mean(numAtts["lstNumTotAtts"]), 2)
-            cIdToAttsData[contentID]["numUsersPart"] = numUsersParticipated
-    sortedData = sorted(cIdToAttsData.items(), key = lambda entry : entry[1]["medianUqAtts"], reverse=True)
-    outputCsv = open("generated_csvs/{}/{}_activity_difficulty.csv".format(zyBookCode, zyBookCode),
+                    userIDtoUser[userID].numTotAtts += 1
+                    userIDtoUser[userID].uqSubs.add(misc_funcs.stripNoiseSub(rawSub))
+        currActivity.numUsersPart = len(userIDtoUser)
+        if currActivity.numUsersPart != 0:   #to avoid div by zero in calc of avg num attempts and bc we don't care about these activities
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Please move comments above lines of code

File: lib/activity_difficulty.py [hunk 13]
```diff
@@ -28,24 +31,23 @@ def generateData(ogCsvData):
...
+            for user in userIDtoUser.values():
+                currActivity.lstNumUqAtts.append(len(user.uqSubs))
+                currActivity.lstNumTotAtts.append(user.numTotAtts)
+            currActivity.medianUqAtts = median(currActivity.lstNumUqAtts)
```
Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Shouldn't this happen in the Activity class? That is, when `lstNumUqAtts` is set, `medianUqAtts` should be automatically computed.

Comment (AlexEdgcomb) [inline lib/activity_difficulty.py]: Same for the below computations.

File: lib/change_code.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    Takes in a dictionary of data for original csv files: {"zyBookCode" : <zyBookCode>, "cIdToPath" : {<contentID1> : <path1>, ...}} and
+    generates a csv per activity with columns
+    
+    User ID  |  Number of Unique, AlteredSubs  |  Most Altered Sub
+
+    Records are sorted in descending order by user's most different sub. Also outputs, per activity, the percentage of students that
+    changed initial code and the average number of unique, altered submissions.
+    '''
+    zyBookCode = ogCsvData["zyBookCode"]
+    for contentID, path in ogCsvData["cIdToPath"].items():
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: Please briefly comment what is happening in this loop

File: lib/change_code.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        defaultCode = html.unescape(responseContentObj["content_resource"]["payload"]["options"]["htmlContents"]).strip()
+        
+        userIdToUser = defaultdict(User)
+        allUsers = set()
+        
+        with open(path, 'r', encoding="utf8") as csvFile:
+            dataTable = csv.reader(csvFile)
+            next(dataTable)
+            next(dataTable)
+            for dateAndTime, userID, isCorrect, rawSub in dataTable:
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: Please briefly comment what's happening in this loop

File: lib/change_code.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                allUsers.add(userID)
+                if rawSub.strip() != defaultCode:
+                    user = userIdToUser[userID]
+                    user.uqSubs.add(rawSub)
+                    lScore = misc_funcs.levenshtein(rawSub, defaultCode)
+                    if lScore > user.highestLScore:
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: Might comment this: `# Store highest Levenshtein score`

File: lib/change_code.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        userIdToUser = defaultdict(User)
+        allUsers = set()
+        
+        with open(path, 'r', encoding="utf8") as csvFile:
+            dataTable = csv.reader(csvFile)
+            next(dataTable)
+            next(dataTable)
+            for dateAndTime, userID, isCorrect, rawSub in dataTable:
+                allUsers.add(userID)
+                if rawSub.strip() != defaultCode:
```
Comment (AlexEdgcomb) [inline lib/change_code.py]: Might comment this: `# Only consider student code that differs from default code`

Comment (AlexEdgcomb) [inline lib/change_code.py]: Although, algorithmically, why ignore such code? Wouldn't the Levenshtein score just be 0?

File: lib/cheating.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+    Function also generates an additional csv across all activities with columns
+
+    User ID  |  Number of Incorrect First Attempts  |  Number of Total Activities Participated
+
+    The latter csv is sorted by number of incorrect first attempts in descending order.
+    '''
+    zyBookCode = ogCsvData["zyBookCode"]
+    userIdToUser = defaultdict(User)
+    for contentID, path in ogCsvData["cIdToPath"].items():
```
Comment (AlexEdgcomb) [inline lib/cheating.py]: Please briefly comment what this loop is doing

File: lib/cheating.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        userIdToFirstIncSub = dict()
+        analyzedUsers = set()
+        with open(path, 'r', encoding="utf8") as csvFile:
+            dataTable = csv.reader(csvFile)
+            next(dataTable)
+            next(dataTable)
+            for dateAndTime, userID, isCorrect, rawSub in dataTable:
```
Comment (AlexEdgcomb) [inline lib/cheating.py]: Please comment what's happening in this loop

File: lib/cheating.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+                if userID not in analyzedUsers:
```
Comment (AlexEdgcomb) [inline lib/cheating.py]: Please combine this and the next if-statement into 1 if-statement. Also, comment when that if-statement will be true

File: lib/user.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
+from collections import defaultdict
+
+#definition of User class
```
Comment (AlexEdgcomb) [inline lib/user.py]: Please document this class a bit more. What's a `User`? What are the internal variables meanings?

File: lib/user_data.py [hunk 22]
```diff
@@ -34,52 +34,58 @@ def generateData(ogCsvData):
...
-        userIdToData[userID]["medNumTotAtts"] = median(numAtts["lstNumTotAtts"])
-    sortedData = sorted(userIdToData.items(), key = lambda entry : entry[1]["medNumUqAtts"], reverse = True)
-    for userID, data in sortedData:
-        writer.writerow([userID, data["medNumUqAtts"], data["avgNumUqAtts"], data["medNumTotAtts"], data["avgNumTotAtts"], data["numActsPart"],
-                         data["numActsCompleted"], data["cIdsUnfinished"]])
+                    if strippedSub not in user.cIdToData[contentID]["uqSubs"]:
+                        activityStats["numUqAtts"] += 1
+                    activityStats["uqSubs"].add(strippedSub)
+                    activityStats["numTotAtts"] += 1
+    for userID, user in userIdToUser.items():
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Please briefly explain what this loop is doing

File: lib/user_data.py [hunk 23]
```diff
@@ -34,52 +34,58 @@ def generateData(ogCsvData):
...
+        user = userIdToUser[userID]
+        activityStats = user.cIdToData[contentID]
+        lstNumUqAtts = []
+        lstNumTotAtts = []
+        for data in user.cIdToData.values():
+            lstNumUqAtts.append(data["numUqAtts"])
+            lstNumTotAtts.append(data["numTotAtts"])
+        user.numActsPart = len(user.cIdToData.keys())
```
Comment (AlexEdgcomb) [inline lib/user_data.py]: Looks like this block of code should be computed in the User class.

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , I made a code review pass.
